File: backend/app/erp/capabilities.py
"""BoHub ERP · roles y permisos — modelo de CAPACIDADES.

Un usuario tiene uno o varios roles (rol principal `User.role` + `User.erp_roles`
opcionales). Cada rol agrupa un conjunto de CAPACIDADES; el permiso efectivo del
usuario es la UNIÓN de las capacidades de todos sus roles. Es la ÚNICA fuente de
verdad de la autorización del ERP: los guards de los endpoints y el frontend
(vía `/api/auth/me`) preguntan por capacidad, no por rol.

Roles operativos del ERP:
  - `comercial`   — ventas: trabaja pedidos NO web de punta a punta, pero no
    cobra en FACTUSOL, no ve pedidos web, ni toca seguimiento/config/integ/
    conciliación. En la Cola SAT solo ve y sube etiqueta.
  - `pedidos`     — ERP Pedidos: todo en pedidos (web y no web) + Cola SAT
    completa + cobros FACTUSOL + seguimiento. No config/integ/conciliación.
  - `sat`         — taller: Cola SAT (preparación/embalado) + envío completo
    (etiqueta/tracking/nº serie/WhiteRIP) + enviar al SAT por email. No
    factura/cobra/crea pedidos; ve pedidos (incl. web) en solo lectura.
  - `admin`       — acceso total + asignar roles.

Reconciliación con lo anterior (documentado en el PR): antes solo había guards
por conjunto de roles (`require_erp_view/edit/admin`). Los roles legacy se mapean:
`manager` → como ERP Pedidos (pierde conciliación, que pasa a admin-only);
`user` → solo lectura del ERP (los comerciales deben pasar al nuevo rol
`comercial`); `viewer` → sin ERP.
"""
from __future__ import annotations

import json
from typing import Any


class Cap:
    """Capacidades del ERP (strings estables; se exponen al frontend)."""

    ACCESS = "erp.access"                    # entrar/ver el ERP
    ORDERS_VIEW_WEB = "erp.orders.view_web"  # ver/actuar sobre pedidos WEB
    ORDERS_CREATE = "erp.orders.create"      # crear pedido manual / desde FACTUSOL
    ORDERS_APPROVE = "erp.orders.approve"    # aprobar pedido
    ORDERS_CANCEL = "erp.orders.cancel"      # anular / restaurar
    SAMPLES_CREATE = "erp.samples.create"    # crear muestra / envío no facturable
    ALBARAN_CREATE = "erp.albaran.create"    # crear albarán en FACTUSOL
    INVOICE_EMIT = "erp.invoice.emit"        # emitir factura en FACTUSOL
    COBRO_REGISTER = "erp.cobro.register"    # registrar cobro en FACTUSOL
    EMAIL_SAT = "erp.email.sat"              # enviar el pedido al taller por email
    EMAIL_CLIENT = "erp.email.client"        # enviar factura/pedido al cliente
    PROFORMAS = "erp.proformas"              # proformas: crear/duplicar/convertir
    DOCUMENTS = "erp.documents"              # documentos FACTUSOL: ver/crear/vincular
    COMPANIES = "erp.companies"              # empresas: crear/vincular/editar/traer
    SAT_VIEW = "erp.sat.view"                # ver la Cola SAT
    SAT_PREPARE = "erp.sat.prepare"          # preparación (empezar/embalar/recoger)
    SAT_SHIPPING = "erp.sat.shipping"        # envío: crear / subir etiqueta
    SAT_TRACKING = "erp.sat.tracking"        # tracking / nº serie / WhiteRIP
    SAT_NO_SHIPPING = "erp.sat.no_shipping"  # «No requiere envío» (masivo)
    SEGUIMIENTO = "erp.seguimiento"          # hoja de seguimiento (ver/editar/Drive)
    CONCILIACION = "erp.conciliacion"        # conciliación bancaria
    CONFIG = "erp.config"                    # configuración ERP
    INTEGRACIONES = "erp.integraciones"      # integraciones (Woo, etc.)
    ROLES_ASSIGN = "admin.roles"             # asignar roles a usuarios


#: Todas las capacidades (lo que tiene el admin).
ALL_CAPS: frozenset[str] = frozenset(
    v for k, v in vars(Cap).items() if not k.startswith("_") and isinstance(v, str)
)

# --- conjuntos por rol -------------------------------------------------------

_COMERCIAL_CAPS: frozenset[str] = frozenset({
    Cap.ACCESS, Cap.ORDERS_CREATE, Cap.ORDERS_APPROVE, Cap.ORDERS_CANCEL,
    Cap.ALBARAN_CREATE, Cap.INVOICE_EMIT, Cap.EMAIL_SAT, Cap.EMAIL_CLIENT,
    Cap.PROFORMAS, Cap.DOCUMENTS, Cap.COMPANIES, Cap.SAT_VIEW, Cap.SAT_SHIPPING,
    Cap.SAMPLES_CREATE,
})
#: ERP Pedidos = comercial + web + cobro + Cola SAT completa + seguimiento.
_PEDIDOS_CAPS: frozenset[str] = _COMERCIAL_CAPS | frozenset({
    Cap.ORDERS_VIEW_WEB, Cap.COBRO_REGISTER, Cap.SAT_PREPARE, Cap.SAT_TRACKING,
    Cap.SAT_NO_SHIPPING, Cap.SEGUIMIENTO,
})
#: ERP SAT = taller + envío/técnicos + enviar al SAT + ver pedidos (incl. web).
#: El taller también DA DE ALTA muestras (además de prepararlas y enviarlas):
#: una muestra no es facturable, así que crearla no requiere ser «oficina».
_SAT_CAPS: frozenset[str] = frozenset({
    Cap.ACCESS, Cap.ORDERS_VIEW_WEB, Cap.SAT_VIEW, Cap.SAT_PREPARE,
    Cap.SAT_SHIPPING, Cap.SAT_TRACKING, Cap.SAT_NO_SHIPPING, Cap.EMAIL_SAT,
    Cap.SAMPLES_CREATE,
})
#: Legacy `user`: solo lectura del ERP (sin edición). Reasignar a `comercial`.
_USER_CAPS: frozenset[str] = frozenset({
    Cap.ACCESS, Cap.ORDERS_VIEW_WEB, Cap.SAT_VIEW,
})

#: Rol → capacidades. `admin` va aparte (ALL_CAPS).
ROLE_CAPABILITIES: dict[str, frozenset[str]] = {
    "comercial": _COMERCIAL_CAPS,
    "pedidos": _PEDIDOS_CAPS,
    "sat": _SAT_CAPS,
    # Legacy — reconciliación:
    "manager": _USER_CAPS,      # como antes: ve el ERP, no edita (CRM-focused)
    "user": _USER_CAPS,         # solo lectura del ERP
    "viewer": frozenset(),      # sin ERP
}

#: Roles OPERATIVOS del ERP que se pueden asignar como rol adicional (multi-rol)
#: y que se ofrecen en /admin/users.
ASSIGNABLE_ERP_ROLES: tuple[str, ...] = ("comercial", "pedidos", "sat")


def _role_value(role: Any) -> str:
    return str(getattr(role, "value", role) or "").strip()


def parse_erp_roles(raw: Any) -> list[str]:
    """Lista de roles operativos válidos guardada en `User.erp_roles` (JSON)."""
    if not raw:
        return []
    try:
        data = json.loads(raw) if isinstance(raw, str) else raw
    except (TypeError, ValueError):
        return []
    if not isinstance(data, list):
        return []
    out: list[str] = []
    for item in data:
        val = _role_value(item)
        if val in ASSIGNABLE_ERP_ROLES and val not in out:
            out.append(val)
    return out


def effective_roles(user: Any) -> set[str]:
    """Roles efectivos del usuario: el principal (`role`) + los operativos
    adicionales (`erp_roles`)."""
    roles = {_role_value(getattr(user, "role", ""))}
    roles |= set(parse_erp_roles(getattr(user, "erp_roles", None)))
    return {r for r in roles if r}


def capabilities_for(user: Any) -> frozenset[str]:
    """Capacidades EFECTIVAS del usuario (unión de las de todos sus roles).
    `admin` tiene todas."""
    roles = effective_roles(user)
    if "admin" in roles:
        return ALL_CAPS
    caps: set[str] = set()
    for role in roles:
        caps |= ROLE_CAPABILITIES.get(role, frozenset())
    return frozenset(caps)


def has_capability(user: Any, cap: str) -> bool:
    """¿El usuario tiene la capacidad `cap`?"""
    roles = effective_roles(user)
    if "admin" in roles:
        return True
    return any(cap in ROLE_CAPABILITIES.get(role, frozenset()) for role in roles)


def can_view_web_orders(user: Any) -> bool:
    """¿Puede ver pedidos de origen web? (Los que no, no ven ni tocan web.)"""
    return has_capability(user, Cap.ORDERS_VIEW_WEB)

File: frontend/src/app/lib/capabilities.ts
import type { Role, User } from "./api";

/** BoHub ERP · roles y permisos — modelo de CAPACIDADES (espejo del backend
 *  `app/erp/capabilities.py`). El frontend gatea qué acciones/menús muestra por
 *  CAPACIDAD, no por rol: el permiso efectivo es la UNIÓN de las capacidades de
 *  todos los roles del usuario (rol principal `role` + `erp_roles`).
 *
 *  `/api/auth/me` ya trae `capabilities` calculado por el backend; cuando está,
 *  se usa tal cual. Para objetos `User` que NO lo traen (p.ej. la lista de
 *  usuarios) se deriva de los roles con el mismo mapa que el backend. En ambos
 *  casos el backend es la fuente de verdad: esto solo esconde/deshabilita en la
 *  UI y cada endpoint vuelve a comprobar la capacidad (403 si falta). */
export const Cap = {
  ACCESS: "erp.access",
  ORDERS_VIEW_WEB: "erp.orders.view_web",
  ORDERS_CREATE: "erp.orders.create",
  ORDERS_APPROVE: "erp.orders.approve",
  ORDERS_CANCEL: "erp.orders.cancel",
  SAMPLES_CREATE: "erp.samples.create",
  ALBARAN_CREATE: "erp.albaran.create",
  INVOICE_EMIT: "erp.invoice.emit",
  COBRO_REGISTER: "erp.cobro.register",
  EMAIL_SAT: "erp.email.sat",
  EMAIL_CLIENT: "erp.email.client",
  PROFORMAS: "erp.proformas",
  DOCUMENTS: "erp.documents",
  COMPANIES: "erp.companies",
  SAT_VIEW: "erp.sat.view",
  SAT_PREPARE: "erp.sat.prepare",
  SAT_SHIPPING: "erp.sat.shipping",
  SAT_TRACKING: "erp.sat.tracking",
  SAT_NO_SHIPPING: "erp.sat.no_shipping",
  SEGUIMIENTO: "erp.seguimiento",
  CONCILIACION: "erp.conciliacion",
  CONFIG: "erp.config",
  INTEGRACIONES: "erp.integraciones",
  ROLES_ASSIGN: "admin.roles",
} as const;

export type Capability = (typeof Cap)[keyof typeof Cap];

/** Todas las capacidades (lo que tiene el admin). */
const ALL_CAPS: ReadonlyArray<string> = Object.values(Cap);

// --- conjuntos por rol (espejo de ROLE_CAPABILITIES del backend) -------------

const COMERCIAL_CAPS: ReadonlyArray<string> = [
  Cap.ACCESS, Cap.ORDERS_CREATE, Cap.ORDERS_APPROVE, Cap.ORDERS_CANCEL,
  Cap.ALBARAN_CREATE, Cap.INVOICE_EMIT, Cap.EMAIL_SAT, Cap.EMAIL_CLIENT,
  Cap.PROFORMAS, Cap.DOCUMENTS, Cap.COMPANIES, Cap.SAT_VIEW, Cap.SAT_SHIPPING,
  Cap.SAMPLES_CREATE,
];
const PEDIDOS_CAPS: ReadonlyArray<string> = [
  ...COMERCIAL_CAPS,
  Cap.ORDERS_VIEW_WEB, Cap.COBRO_REGISTER, Cap.SAT_PREPARE, Cap.SAT_TRACKING,
  Cap.SAT_NO_SHIPPING, Cap.SEGUIMIENTO,
];
const SAT_CAPS: ReadonlyArray<string> = [
  Cap.ACCESS, Cap.ORDERS_VIEW_WEB, Cap.SAT_VIEW, Cap.SAT_PREPARE,
  Cap.SAT_SHIPPING, Cap.SAT_TRACKING, Cap.SAT_NO_SHIPPING, Cap.EMAIL_SAT,
  // El taller también DA DE ALTA muestras, no solo las prepara.
  Cap.SAMPLES_CREATE,
];
/** Legacy `manager`/`user`: solo lectura del ERP (reasignar a `comercial`). */
const USER_CAPS: ReadonlyArray<string> = [
  Cap.ACCESS, Cap.ORDERS_VIEW_WEB, Cap.SAT_VIEW,
];

const ROLE_CAPABILITIES: Record<Role, ReadonlyArray<string>> = {
  admin: ALL_CAPS,
  comercial: COMERCIAL_CAPS,
  pedidos: PEDIDOS_CAPS,
  sat: SAT_CAPS,
  manager: USER_CAPS,
  user: USER_CAPS,
  viewer: [],
};

/** Roles operativos del ERP que se pueden asignar como rol adicional (multi-rol)
 *  en /admin/users. Espejo de `ASSIGNABLE_ERP_ROLES` del backend. */
export const ASSIGNABLE_ERP_ROLES: ReadonlyArray<Role> = ["comercial", "pedidos", "sat"];

type CapUser = Pick<User, "role"> & {
  capabilities?: string[] | null;
  erp_roles?: string[] | null;
};

/** Capacidades EFECTIVAS del usuario. Prefiere el array `capabilities` de
 *  `/api/auth/me`; si no viene, lo deriva de los roles (principal + `erp_roles`)
 *  con el mismo mapa que el backend. */
export function effectiveCapabilities(
  user: CapUser | null | undefined,
): ReadonlyArray<string> {
  if (!user) return [];
  // `/api/auth/me` trae el array ya calculado (incluido `[]` para viewer): se
  // usa tal cual. Un `User` sin él (lista de usuarios) se deriva de los roles.
  if (Array.isArray(user.capabilities)) return user.capabilities;
  const roles = new Set<Role>([user.role]);
  for (const extra of user.erp_roles ?? []) {
    if ((ASSIGNABLE_ERP_ROLES as readonly string[]).includes(extra)) {
      roles.add(extra as Role);
    }
  }
  if (roles.has("admin")) return ALL_CAPS;
  const caps = new Set<string>();
  for (const role of roles) {
    for (const cap of ROLE_CAPABILITIES[role] ?? []) caps.add(cap);
  }
  return [...caps];
}

/** ¿El usuario tiene la capacidad `cap`? (unión de las de todos sus roles). */
export function can(
  user: CapUser | null | undefined,
  cap: Capability | string,
): boolean {
  if (!user) return false;
  if (user.role === "admin") return true;
  return effectiveCapabilities(user).includes(cap);
}

/** ¿Puede ver/actuar sobre pedidos de origen web (WooCommerce)? El Comercial no:
 *  los pedidos web quedan fuera de sus listas y su ficha le devuelve 403. */
export function canViewWebOrders(user: CapUser | null | undefined): boolean {
  return can(user, Cap.ORDERS_VIEW_WEB);
}
